stdpopsim/slim_engine.py

```diff
@@ -0,0 +1,663 @@
+    defineConstant("trees_file", "$trees_file");
+    defineConstant("check_coalescence", $check_coalescence);
+"""
+
+_slim_lower = """
+    defineConstant("N", asInteger(_N/Q));
+
+    initializeTreeSeq(checkCoalescence=check_coalescence);
+    initializeMutationRate(mutation_rate);
+    initializeMutationType("m1", 0.5, "f", 0);
+    initializeGenomicElementType("g1", m1, 1.0);
+    initializeGenomicElement(g1, 0, chromosome_length-1);
+    initializeRecombinationRate(recombination_rate);
+}
+
+function (void)dbg(string$ s, [integer$ debug_level = 2]) {
+    if (verbosity >= debug_level)
+        catn(sim.generation + ": " + s);
+}
+
+// Return the number of generations that separate t0 and t1.
+function (integer)gdiff(numeric$ t0, numeric t1) {
+    return asInteger(round((t0-t1)/generation_time/Q));
+}
+
+// Output tree sequence file and end the simulation.
+function (void)end(void) {
+    sim.treeSeqOutput(trees_file);
+    sim.simulationFinished();
+}
+
+// Create initial populations and migration rates.
+1 {
+    // Create initial populations.
+    for (i in 0:(num_populations-1)) {
+        if (N[0,i] > 0) {
+            dbg("sim.addSubpop("+i+", "+N[0,i]+");");
+            sim.addSubpop(i, N[0,i]);
+        }
+    }
+
+    // Migration rates.
+    i = 0;
+    for (j in 0:(num_populations-1)) {
+        for (k in 0:(num_populations-1)) {
+            if (j==k | N[i,j] == 0 | N[i,k] == 0)
+                next;
+
+            m = migration_matrices[k,j,i];
+            p = sim.subpopulations[j];
+            dbg("p"+j+".setMigrationRates("+k+", "+m+");");
+            p.setMigrationRates(k, m);
+        }
+    }
+}
+
+// Check if burn in has completed.
+1 {
+    if (!check_coalescence) {
+        setup();
+        return;
+    }
+
+    if (sim.treeSeqCoalesced()) {
+        /*
+         * All current generation individuals now have a common ancestor
+         * born after the start of our simulation.  But the genealogy at this
+         * first coalescence is not a good representative of an average
+         * genealogy, as the TMRCA is biased low.  To understand why,
+         * consider a haploid population with constant N=2.  At first
+         * coalescence, the TMRCA is always one generation, yet clearly
+         * genealogies with longer TMRCA's are possible for this population.
+         *
+         * To obtain a burn-in genealogy drawn from the full distribution
+         * of possible genealogies, we may continue the simulation forwards
+         * until this ergodic process converges.  Unfortunately, it's not
+         * clear how many more generations are required for convergence,
+         * and this is likely dependent on demography, and non-neutral
+         * processes (if any).  10*N appears to be reasonable for a single
+         * neutrally evolving population with constant N.
+         */
+        N_max = max(N[0,0:(num_populations-1)]);
+        g = sim.generation + 10 * N_max;
```

Comment:
I think we are going to want to have the burn-in period as a parameter available through the API, rather than be fixed in this way

Reply:
Yeah, that's a good idea. At the moment this code waits for all individuals to have a common ancestor after the start of the simulation (via `sim.treeSeqCoalesced()`), and *then* waits `10*N` more generations. So should we:
1. Wait until coalescence and have the burn-in period parameter change the `10*N`, or
2. remove the coalescence check altogether, potentially leaving genealogies without a single root. (Note this burn-in will only be used when recapitation is disabled).
3. Is it ok to have the parameter in units of `N` generations? (Where N is the size of the largest population that exists at the start of the simulation---no model currently has more than one population).